core/src/TBB/Kokkos_TBB_Task.hpp

```diff
@@ -0,0 +1,453 @@
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_TBB_TASK_HPP
+#define KOKKOS_TBB_TASK_HPP
+
+#include <Kokkos_Macros.hpp>
+#if defined(KOKKOS_ENABLE_TBB) && defined(KOKKOS_ENABLE_TASKDAG)
+
+#include <Kokkos_TaskScheduler_fwd.hpp>
+
+#include <Kokkos_TBB.hpp>
+
+#include <type_traits>
+
+//----------------------------------------------------------------------------
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+namespace Impl {
+
+  template<class SchedulerType, class QueueType>
+  class TBBTask : public tbb::task {
+    using scheduler_type = SchedulerType;
+    using task_base_type = typename scheduler_type::task_base_type;
+
+  public:
+    TBBTask(OptionalRef<task_base_type> taskToRun,
+         tbb::empty_task* iWaitingTask,
+         scheduler_type iScheduler):
+      task(std::move(taskToRun)),
+      waitingTask(iWaitingTask),
+      scheduler(iScheduler) {
+    }
+
+    tbb::task* execute() {
+      const int num_worker_threads = Kokkos::Experimental::TBB::concurrency();
+
+      // NOTE: This implementation has been simplified based on the
+      // assumption that team_size = 1. The TBB backend currently only
+      // supports a team size of 1.
+      std::size_t t = Kokkos::Experimental::TBB::impl_hardware_thread_id();
+
+      intptr_t buffer[512/sizeof(intptr_t)];
+      TBBTeamMember member(TeamPolicyInternal<Kokkos::Experimental::TBB>(
+                                                                         Kokkos::Experimental::TBB(), num_worker_threads, 1),
+                           0, t, buffer, 512);
+
+      using member_type =
+        TaskTeamMemberAdapter<Kokkos::Impl::TBBTeamMember, scheduler_type>;
+      
+      member_type single_exec(scheduler, member);
+      member_type &team_exec = single_exec;
+      
+      auto &team_scheduler = team_exec.scheduler();
+      task->as_runnable_task().run(single_exec);
+
+      auto &queue = scheduler.queue();
+
+      //This is where any tasks waiting on this one will be scheduled
+
+      //NOTE: does team_scheduler.team_scheduler_info have to be the same as the 
+      // value used when calling queue.pop_ready_task? If so, then can't do it this way
+      queue.complete((*std::move(task)).as_runnable_task(),
```

Comment:
What does `*std::move(task)` do?

Reply:
I cribed it from here
https://github.com/Dr15Jones/kokkos/blob/tbb_hack/core/src/HPX/Kokkos_HPX_Task.hpp#L126